### PR TITLE
Requirements now depend on python version

### PR DESCRIPTION
**Pull request by potiuk**
Generate a separate requirements.txt file for each Python Version we support.

---

Make sure to mark the boxes below before creating PR: [x]

- [x] Description above provides context of the change
- [x] Unit tests coverage for changes (not needed for documentation changes)
- [x] Commits follow "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)"
- [x] Relevant documentation is updated including usage instructions.
- [x] I will engage committers as explained in [Contribution Workflow Example](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#contribution-workflow-example).

---
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/master/UPDATING.md).
Read the [Pull Request Guidelines](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#pull-request-guidelines) for more information.

### CODE REVIEW

**Comment by potiuk** [comment]
IT turned out that requirements are slightly different for different versions of python (especially in 1.10) so I had to add different version of requirements. At the same time I removed -pinned version. It's not really needed we should be able to easily install airflow now with one command:

```
  pip install apache-airflow[gcp]==1.10.10 \
      --constraint https://raw.githubusercontent.com/apache/airflow/1.10.10/requirements/requirements-python3.7.txt

```

File: .travis.yml [hunk 1]
```diff
@@ -41,24 +41,24 @@ jobs:
       stage: pre-test
       script: ./scripts/ci/ci_run_all_static_checks.sh
       env: >-
-        PYTHON_VERSION=3.6
+        PYTHON_SHORT_VERSION=3.6
```
Comment (kaxil) [inline .travis.yml]: Is there a reason we renamed PYTHON_VERSION to PYTHON_SHORT_VERSION?

Comment (potiuk) [reply .travis.yml]: Yes. PYTHON_VERSION is set by default by the python base image and it is the full version  (PYTHON_VERSION=3.7.9). It got so confusing that I decided I finally have to separate those two.

**Comment by potiuk** [comment]
BTW. I am going to re-write some parts of it. I came to the conclusion that generating requirements is too big of a hassle for pre-commits (especially that we need them in different versions) so I will do a separate breeze command (and script for those not using Breeze) so that whenever you modify setup.py and new requirements are needed you will be notified via CI rather than pre-commit. I think it's much less impact on the development workflow.

Bear with me :).

**Comment by potiuk** [comment]
@kaxil @ashb @turbaszek @mik-laj @dimberman  -> We have much nicer version now. Requirement generation is now checked only as travis build (separately for each python version) rather than pre-commit and you can esily manually generate the requirements with breeze (examples in doc). Also when you modify setup.py in the way that it will require regeneration of the reuqirements, the build in CI will fail, show you with nice, coloured diff what is added and will ask you to run `breeze generate-requirements --python x.y` as needed. 

Pls review. That's the last thing to cherry-pick and test prod image in airflow 1.10.10

File: .travis.yml (hunk 1 above)
Comment (kaxil) [reply .travis.yml]: Sounds good

File: CONTRIBUTING.rst [hunk 2]
```diff
@@ -341,31 +341,59 @@ be open to allow several different libraries with the same requirements to be in
...
+------------------------
 
-* we have ``requirements.txt`` file generated automatically based on the set of all latest working
-  and tested requirement versions. You can also use that file as a constraints file when installing
-  apache airflow - either from the sources ``pip install -e . --constraint requirements.txt`` or
-  from the pypi package ``pip install apache-airflow --constraint requirements.txt``. Note that
-  this will also work with extras for example ``pip install .[gcp] --constraint requirements.txt`` or
-  ``pip install apache-airflow[gcp] --constraint requirements.txt``
+By default when you install ``apache-airflow`` package - the dependencies are as open as possible while
+still allowing the apache-airflow to install. This means that 'apache-airflow' package might fail to
```
Comment (kaxil) [inline CONTRIBUTING.rst]:
```suggestion
still allowing the apache-airflow package to install. This means that 'apache-airflow' package might fail to
```

File: CONTRIBUTING.rst [hunk 3]
```diff
@@ -341,31 +341,59 @@ be open to allow several different libraries with the same requirements to be in
...
+install in case a direct or transitive dependency is released that breaks the installation. In such case
+when installing ``apache-airflow``, you might need to provide additional constraints (for
+example ``pip install apache-airflow==1.10.2 Werkzeug<1.0.0``)
 
-The ``requirements.txt`` file should be updated automatically via pre-commit whenever you update dependencies
-It reflects the current set of dependencies installed in the CI image of Apache Airflow.
-The same set of requirements will be used to produce the production image.
+However we now havehave ``requirements-python<PYTHON_SHORT_VERSION>.txt`` file generated
```
Comment (kaxil) [inline CONTRIBUTING.rst]:
```suggestion
However, we now have ``requirements-python<PYTHON_SHORT_VERSION>.txt`` file generated
```

File: CONTRIBUTING.rst [hunk 4]
```diff
@@ -341,31 +341,59 @@ be open to allow several different libraries with the same requirements to be in
...
+automatically and committd in the requirements folder based on the set of all latest working and tested
```
Comment (kaxil) [inline CONTRIBUTING.rst]:
```suggestion
automatically and committed in the requirements folder based on the set of all the latest working and tested
```

File: .travis.yml (hunk 1 above)
Comment (mik-laj) [reply .travis.yml]: What do you think about renaming to ``PYTHON_MINOR_VERSION``?

File: CONTRIBUTING.rst [hunk 5]
```diff
@@ -341,31 +341,59 @@ be open to allow several different libraries with the same requirements to be in
...
+requirement versions. Those requirement.txt files can be used as constraints file when installing apache
```
Comment (kaxil) [inline CONTRIBUTING.rst]:
```suggestion
requirement versions. Those ``requirements.txt`` files can be used as constraints file when installing apache
```

File: CONTRIBUTING.rst [hunk 6]
```diff
@@ -341,31 +341,59 @@ be open to allow several different libraries with the same requirements to be in
...
+airflow - either from the sources ``pip install -e . --constraint requirements/requirements-python3.7.txt``
+or from the pypi package ``pip install apache-airflow --constraint requirements/requirements-python3.7.txt``.
+This work also with extras - ``pip install .[gcp] --constraint requirements/requirements-python3.7.txt``.
```
Comment (kaxil) [inline CONTRIBUTING.rst]:
```suggestion
This also works with extras - ``pip install .[gcp] --constraint requirements/requirements-python3.7.txt``.
```

File: CONTRIBUTING.rst [hunk 7]
```diff
@@ -341,31 +341,59 @@ be open to allow several different libraries with the same requirements to be in
...
+It is also possible to use constraints directly from github (note that it only works starting from
+airflow 1.10.10 version:
```
Comment (kaxil) [inline CONTRIBUTING.rst]:
```suggestion
Airflow 1.10.10 version:
```

File: CONTRIBUTING.rst [hunk 8]
```diff
@@ -341,31 +341,59 @@ be open to allow several different libraries with the same requirements to be in
...
-``./scripts/ci/ci_generate_requirements.sh``.
-This will try to regenerate the requirements.txt file with the latest requirements matching
-the setup.py constraints.
+.. code-block:: bash
+
+  pip install apache-airflow[gcp]==1.10.10 \
+      --constraint https://raw.githubusercontent.com/apache/airflow/1.10.10/requirements/requirements-python3.7.txt
+
+There are different set of fixed requirements for different python major/minor versions and you should
+use the right python version file for the right python version.
```
Comment (kaxil) [inline CONTRIBUTING.rst]:
```suggestion
use the right requirements file for the right python version.
```

File: CONTRIBUTING.rst [hunk 9]
```diff
@@ -341,31 +341,59 @@ be open to allow several different libraries with the same requirements to be in
...
+
+The ``requirements-python<PYTHON_SHORT_VERSION>.txt`` file MUST be regenerated every time after setup.py
```
Comment (kaxil) [inline CONTRIBUTING.rst]:
```suggestion
The ``requirements-python<PYTHON_SHORT_VERSION>.txt`` file MUST be regenerated every time after ``setup.py``
```

File: CONTRIBUTING.rst [hunk 10]
```diff
@@ -341,31 +341,59 @@ be open to allow several different libraries with the same requirements to be in
...
+is updated. This is checked automatically in Travis CI build there are separate jobs for each
```
Comment (kaxil) [inline CONTRIBUTING.rst]:
```suggestion
is updated. This is checked automatically in Travis CI build. There are separate jobs for each
```

File: CONTRIBUTING.rst [hunk 11]
```diff
@@ -341,31 +341,59 @@ be open to allow several different libraries with the same requirements to be in
...
+python version that checks if the requirements should be updated.
+
+If they are not you should regenerate the requirements locally using one of the methods described below.
```
Comment (kaxil) [inline CONTRIBUTING.rst]:
```suggestion
If they are not updated, you should regenerate the requirements locally using one of the methods described below.
```

File: CONTRIBUTING.rst [hunk 12]
```diff
@@ -341,31 +341,59 @@ be open to allow several different libraries with the same requirements to be in
...
+
+Generating requirement files
+----------------------------
+
+This shuould be done every time after you modify setup.py file. You can generate requirement files
```
Comment (kaxil) [inline CONTRIBUTING.rst]:
```suggestion
This should be done every time after you modify setup.py file. You can generate requirement files
```

File: CONTRIBUTING.rst [hunk 13]
```diff
@@ -341,31 +341,59 @@ be open to allow several different libraries with the same requirements to be in
 The problem is that Apache Airflow is a bit of both - application to install and library to be used when
 you are developing your own operators and DAGs.
 
-This - seemingly unsolvable - puzzle is solved as follows:
+This - seemingly unsolvable - puzzle is solved by having pinned requirement files. Those are available
+as of airflow 1.10.10.
```
Comment (mik-laj) [inline CONTRIBUTING.rst:345]:
```suggestion
as of Airflow 1.10.10.
```

File: scripts/ci/_utils.sh [hunk 14]
```diff
@@ -530,7 +531,7 @@ function confirm_image_rebuild() {
         # Make sure to use /dev/tty first rather than stdin/stdout when available - this way confirm
         # will works also in case of pre-commits (git does not pass stdin/stdout to pre-commit hooks)
         # shellcheck disable=SC2094
-        "${AIRFLOW_SOURCES}/confirm" "${ACTION} image ${THE_IMAGE_TYPE} (might take some time)" \
+        "${AIRFLOW_SOURCES}/confirm" "${ACTION} image ${THE_IMAGE_TYPE}-${PYTHON_SHORT_VERSION}" \
```
Comment (kaxil) [inline scripts/ci/_utils.sh]:
```suggestion
        "${AIRFLOW_SOURCES}/confirm" "${ACTION} image ${THE_IMAGE_TYPE}-python ${PYTHON_SHORT_VERSION}" \
```

?? not sure

File: scripts/ci/docker-compose/base.yml [hunk 15]
```diff
@@ -59,6 +59,8 @@ services:
       - DB_RESET
       - ENABLED_SYSTEMS
       - RUN_SYSTEM_TESTS
+      - DEFAULT_PYTHON_SHORT_VERSION
+      - PYTHON_SHORT_VERSION
```
Comment (kaxil) [inline scripts/ci/docker-compose/base.yml]: Do we need both?

File: .travis.yml (hunk 1 above)
Comment (potiuk) [reply .travis.yml]: PYTHON_MAJOR_MINOR_VERSION is better

File: CONTRIBUTING.rst (hunk 2 above)
Comment (potiuk) [reply CONTRIBUTING.rst]: added

File: CONTRIBUTING.rst (hunk 4 above)
Comment (potiuk) [reply CONTRIBUTING.rst]: Done

File: CONTRIBUTING.rst (hunk 3 above)
Comment (potiuk) [reply CONTRIBUTING.rst]: Fixed

File: CONTRIBUTING.rst (hunk 5 above)
Comment (potiuk) [reply CONTRIBUTING.rst]: Fixed

File: CONTRIBUTING.rst (hunk 7 above)
Comment (potiuk) [reply CONTRIBUTING.rst]: rephrased

File: CONTRIBUTING.rst (hunk 9 above)
Comment (potiuk) [reply CONTRIBUTING.rst]: Fixed

File: CONTRIBUTING.rst (hunk 10 above)
Comment (potiuk) [reply CONTRIBUTING.rst]: Fixed,

File: CONTRIBUTING.rst (hunk 12 above)
Comment (potiuk) [reply CONTRIBUTING.rst]: Fixed

File: CONTRIBUTING.rst (hunk 11 above)
Comment (potiuk) [reply CONTRIBUTING.rst]: Fixed

File: CONTRIBUTING.rst (hunk 13 above)
Comment (potiuk) [reply CONTRIBUTING.rst:345]: Fixed

File: scripts/ci/_utils.sh (hunk 14 above)
Comment (potiuk) [reply scripts/ci/_utils.sh]: Yep.

File: scripts/ci/docker-compose/base.yml (hunk 15 above)
Comment (potiuk) [reply scripts/ci/docker-compose/base.yml]: Not any more :). It was only needed to generate requirements.txt file in the main repo but I removed it - it was nice for IDE's but very redundant and confusing.

File: .travis.yml [hunk 16]
```diff
@@ -41,24 +41,24 @@ jobs:
       stage: pre-test
       script: ./scripts/ci/ci_run_all_static_checks.sh
       env: >-
-        PYTHON_VERSION=3.6
+        PYTHON_MAJOR_MINOR_VERSION=3.6
```
Comment (turbaszek) [inline .travis.yml:44]: Why such strange name?

**Review by mik-laj — APPROVED** [review]

File: Dockerfile.ci [hunk 17]
```diff
@@ -40,14 +43,19 @@ ARG DEPENDENCIES_EPOCH_NUMBER="2"
...
     && apt-get install -y --no-install-recommends \
            curl \
            gnupg2 \
+           netselect-apt \
     && apt-get autoremove -yqq --purge \
     && apt-get clean \
     && rm -rf /var/lib/apt/lists/*
 
+# Choose faster mirror to download dependencies
+RUN netselect-apt -o /etc/apt/sources.list
```
Comment (mik-laj) [inline Dockerfile.ci]: +1

File: .travis.yml (hunk 16 above)
Comment (potiuk) [reply .travis.yml:44]: It's not strange. It's MAJOR+MINOR version. PYTHON_VERSION is by default set in this image to 3.6.9 (so full PYTHON_VERSION) and what we really need here just MAJOR/MINOR.

Comment (potiuk) [reply .travis.yml:44]: Are you OK with it @turbaszek ?

**Comment by potiuk** [comment]
OK. fixed separate requirements for different python versions is green in CI @kaxil @mik-laj -> do you want to take a look ?

**Review by turbaszek — APPROVED** [review]

**Pull request merged**